msg_truncate.go

```diff
@@ -0,0 +1,106 @@
+package dns
+
+// Truncate ensures the reply message will fit into the requested buffer
+// size by removing records that exceed the requested size.
+//
+// It will first check if the reply fits without compression and then with
+// compression. If it won't fit with compression, Scrub then walks the
+// record adding as many records as possible without exceeding the
+// requested buffer size.
+//
+// The TC bit will be set if any answer records were excluded from the
+// message. This indicates to that the client should retry over TCP.
+//
+// The appropriate buffer size can be retrieved from the requests OPT
+// record, if present, and is transport specific otherwise. dns.MinMsgSize
+// should be used for UDP requests without an OPT record, and
+// dns.MaxMsgSize for TCP requests without an OPT record.
+func (dns *Msg) Truncate(size int) {
+	if dns.IsTsig() != nil {
+		// To simplify this implementation, we don't perform
+		// truncation on responses with a TSIG record.
+		return
+	}
+
+	// RFC 6891 mandates that the payload size in an OPT record
+	// less than 512 bytes must be treated as equal to 512 bytes.
+	//
+	// For ease of use, we impose that restriction here.
+	if size < 512 {
+		size = 512
+	}
+
+	l := msgLenWithCompressionMap(dns, nil) // uncompressed length
+	if l <= size {
+		// Don't waste effort compressing this message.
+		dns.Compress = false
+		return
+	}
+
+	dns.Compress = true
+
+	edns0 := dns.popEdns0()
+	if edns0 != nil {
+		// Account for the OPT record that gets added at the end,
+		// by subtracting that length from our budget.
+		//
+		// The EDNS(0) OPT record must have the root domain and
+		// it's length is thus unaffected by compression.
+		size -= Len(edns0)
+	}
+
+	compression := make(map[string]struct{})
+
+	l = 12 // Message header is always 12 bytes
+	for _, r := range dns.Question {
+		l += r.len(l, compression)
+	}
+
+	var numAnswer int
+	if l < size {
+		l, numAnswer = truncateLoop(dns.Answer, size, l, compression)
+	}
+
+	var numNS int
```

Comment:
truncating NS section... I wonder if that is a) ever needed [yes it can be artificially inflated] b) described in the RFCs somewhere?

Reply:
I don't really think there is any other way to do this. Because the compression map is, and must, be created incrementally, it's not possible to truncate the answer section but not the NS section and still have it fit. You'd either drop far too many answer records, or always exceed the size limit.

You have to walk the message record-by-record and answers go before NS, the only way to not-truncate the NS section, while still being accurate, would be the costly approach in CoreDNS.

That being said, I'll scour the RFCs and see if I can find any guidance here.

Reply:
One option if you don't think the authority section should/can be truncated is to remove the uncompressed length of the authority section from the budget and include them all. This would mean that more answer records are excluded than is actually required.

Reply:
[RFC 1035](https://tools.ietf.org/html/rfc1035#section-6.2):
> When a response is so long that truncation is required, the truncation should start at the end of the response and work forward in the datagram.  Thus if there is any data for the authority section, the answer section is guaranteed to be unique.

There are definitely clients that would break if a truncated message included a section after an incomplete one.

Reply:
@gibson042 Thank you for that quote, that’s what I had been thinking. The approach in this PR is correct.